multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/CloudControllerClientProvider.java

```diff
@@ -123,28 +112,19 @@ private OAuth2AccessTokenWithAdditionalInfo getValidToken(String userName) {
         return token;
     }
 
-    private CloudControllerClient getClientFromCache(String userName, String org, String space) {
-        return getClientFromCache(userName, org, space, null);
-    }
-
-    private CloudControllerClient getClientFromCache(String userName, String org, String space, String processId) {
-        String key = getKey(userName, org, space);
-        CloudControllerClient client = clients.get(key);
-        if (client == null) {
-            client = clientFactory.createClient(getValidToken(userName), org, space);
-            if (processId != null) {
-                clients.put(key, client);
-            }
-        }
-        return client;
+    private CloudControllerClient getClientFromCache(String userName, String org, String space, String correlationId) {
+        String key = getKey(userName, org, space, correlationId);
+        return clients.computeIfAbsent(key, k -> clientFactory.createClient(getValidToken(userName), org, space, correlationId));
     }
 
-    private CloudControllerClient getClientFromCache(String userName, String spaceId) {
-        String key = getKey(userName, spaceId);
-        return clients.computeIfAbsent(key, k -> clientFactory.createClient(getValidToken(userName), spaceId));
+    private CloudControllerClient getClientFromCache(String userName, String spaceId, String correlationId) {
+        String key = getKey(userName, spaceId, correlationId);
+        return clients.computeIfAbsent(key, k -> clientFactory.createClient(getValidToken(userName), spaceId, correlationId));
     }
 
     private String getKey(String... args) {
```

Comment:
What is the advantage of having correlation id in the key at all?
I see 2 disadvantages:
1.  In this way the key become non-deterministic. E.g:
(user_X, my-guid, null) AND (user_X, null, my-guid) will produce the same result 
2. the cache could have more elements, potentially harder to manage

Reply:
Because the client can be created with correlation id or not - imagine the following example:
a user is performing parallel `cf deploy`s in the same space, but for different mta. For each of the deployments, a client will be cached with the correlation id of the process. If the correlation id is not taken into account when managing the clients the might be a process, which uses a client with different correlation id. In fact for a space-user there will be only one client and therefore this feature becomes useless.

Reply:
This is the case with the cached token so far and there are not problems.

Reply:
Currently cache is based on username,org,space or username and spaceguid. Now if we want to tag requests for single process, we need to add in cache CloudControllerClient for every single process, not per space. You can set additional http headers when you create cloudfoundry client from os project.